c++/src/kj/compat/http.c++

```diff
@@ -3489,6 +3489,338 @@ WebSocketPipe newWebSocketPipe() {
 
 // =======================================================================================
 
+namespace _ { // private implementation details
+
+kj::ArrayPtr<const char> splitNext(StringPtr& cursor, char delimiter) {
+  // Consumes and returns the next item in a delimited list.
+  //
+  // If a delimiter is found:
+  //  - `input` is updated to point to the rest of the string after the delimiter.
+  //  - The text before the delimiter is returned.
+  // If no delimiter is found:
+  //  - `input` is updated to an empty string.
+  //  - The text that had been in `input` is returned.
+  //
+  // (It's up to the caller to stop the loop once `input` is empty.)
+  KJ_IF_MAYBE(index, cursor.findFirst(delimiter)) {
+    auto part = cursor.slice(0, *index);
+    cursor = cursor.slice(*index + 1);
+    return part;
+  }
+  auto result = cursor.asArray();
+  cursor = nullptr;
+
+  return result;
+}
+
+void stripLeadingSpace(ArrayPtr<const char>& str) {
+  // Removes any leading spaces from `str`, modifying it in-place.
+  while (str.size() > 0 && (str[0] == ' ' || str[0] == '\t')) {
+    str = str.slice(1, str.size());
+  }
+}
+
+kj::Vector<kj::String> splitPartsWithCondition(StringPtr cursor, char delim, StringPtr condition) {
+  // Given a string `input` and a delimiter `delim`, split the string into a vector of substrings,
+  // separated by the delimiter. Note that leading whitespace is stripped from each element.
+  //
+  // Optionally, the caller may pass a `condition` string, requesting that each substring starts
+  // with the value inside `condition`. This allows for filtering in 1 pass, rather than 2.
+  kj::Vector<kj::String> parts;
+
+  while (cursor.size() != 0) {
+    auto part = splitNext(cursor, delim);
+    stripLeadingSpace(part);
+
+    if (part.startsWith(condition.asArray())) {
+      parts.add(kj::str(part));
+    }
+  }
+
+  return parts;
+}
+
+kj::Array<keyMaybeVal> toKeysAndVals(const kj::ArrayPtr<kj::String>& params) {
+  // Given a collection of parameters (a single offer), parse the parameters into <key, MaybeValue>
+  // pairs. If the parameter contains an `=`, we set the `key` to everything before, and the `value`
+  // to everything after. Otherwise, we set the `key` to be the entire parameter.
+  auto result = kj::heapArray<keyMaybeVal>(params.size());
+  size_t count = 0;
+  for (const auto& param : params) {
+    kj::ArrayPtr<const char> key;
+    kj::Maybe<kj::ArrayPtr<const char>> value;
+
+    KJ_IF_MAYBE(index, param.findFirst('=')) {
+      // Found '=' so we have a value.
+      key = param.slice(0, *index);
+      value = param.slice(*index + 1, param.size());
+    } else {
+      key = param.asArray();
+    }
+
+    result[count] = keyMaybeVal{.key = kj::mv(key), .val = kj::mv(value)};
```

Comment:
https://github.com/capnproto/capnproto/pull/1480/commits/ca8bd4bc78b1f976aa9257c96b28341721f1e34e

Am I correct in expecting `kj::heapArray<KeyMaybeVal>(params.size());` to default initialize each element in the heap allocated array? Judging by the method body of the constructor and this comment in the body of `heapArrayBuilder`:
```
  // Like `heapArray<T>()` but does not default-construct the elements.  You must construct them
  // manually by calling `add()`.
```
I suspect it's fine to assign to the elements of the struct without explicitly initializing them.

Reply:
Yes. Alternatively, you could keep the syntax you had before but just remove `.key =` and `.val =`, so the properties are matched up by order rather than by name.